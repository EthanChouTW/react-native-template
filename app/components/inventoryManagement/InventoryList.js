import React from 'react';
import PropTypes from 'prop-types';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  Text,
  View,
  TouchableOpacity
} from 'react-native';
import Locales, { currencyPrice } from '../../locales';
import Theme from '../../utils/styleCollection';
import { PRODUCT_STATUS_AVAILABLE } from '../../actions/productActions';

const styles = StyleSheet.create({
  constainer: {
    flex: 1,
    paddingVertical: 300
  },
  listContainer: {
    backgroundColor: Theme.hbGray,
    paddingVertical: 8,
    paddingHorizontal: 14
  },
  card: {
    shadowColor: 'black',
    shadowRadius: 1,
    elevation: 2,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.6,
    justifyContent: 'center',
    backgroundColor: 'white',
    margin: 5
  },
  statusButton: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 20,
    borderTopWidth: 1,
    borderColor: Theme.hbGray
  },
  availableText: {
    textAlign: 'center',
    color: Theme.algaeGreenColor
  },
  outOfStockText: {
    textAlign: 'center',
    color: Theme.hbRed
  },
  priceText: {
    paddingHorizontal: 10,
    paddingVertical: 10,
    color: Theme.hbGray
  },
  titleText: {
    paddingTop: 10,
    paddingHorizontal: 10
  }
});

const InventoryList = props => {
  const renderItem = element => (
    <View key={element.item.id} style={styles.card}>
      <Text style={styles.titleText}>{element.item.title}</Text>

      <Text style={styles.priceText}>
        {`${currencyPrice(element.item.price, element.item.currency)}`}
      </Text>
      <TouchableOpacity
        style={styles.statusButton}
        onPress={() => {
          props.onUpdateProductStatus(element.item.id);
        }}
      >
        {props.isChangingStatus === true && props.changingProductId === element.item.id ? (
          <ActivityIndicator animating color={Theme.algaeGreenColor} size="small" />
        ) : (
          <Text
            style={
              element.item.status === PRODUCT_STATUS_AVAILABLE ? (
                styles.availableText
              ) : (
                styles.outOfStockText
              )
            }
          >
            {element.item.status === PRODUCT_STATUS_AVAILABLE ? (
              Locales.t('available').toUpperCase()
            ) : (
              Locales.t('lbl_out_of_stock').toUpperCase()
            )}
          </Text>
        )}
      </TouchableOpacity>
    </View>
  );
  return (
    <FlatList
      removeClippedSubviews={false}
      onRefresh={props.onRefresh}
      refreshing={props.refreshing}
      data={props.data}
      extraData={props.extraData}
      renderItem={renderItem}
      keyExtractor={product => product.id}
      contentContainerStyle={styles.listContainer}
      onEndReached={props.onEndReached}
      onEndReachedThreshold={5}
    />
  );
};

InventoryList.propTypes = {
  onRefresh: PropTypes.func,
  refreshing: PropTypes.bool,
  isChangingStatus: PropTypes.bool,
  changingProductId: PropTypes.number,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      currency: PropTypes.string
    })
  ),
  extraData: PropTypes.shape({
    data: PropTypes.array
  }),
  onEndReached: PropTypes.func,
  onUpdateProductStatus: PropTypes.func
};

export default InventoryList;
