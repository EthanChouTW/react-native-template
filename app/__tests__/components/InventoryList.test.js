import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import InventoryList from '../../components/inventoryManagement/InventoryList';

describe('UserProfileMenuComponent', () => {
  const props = {
    onRefresh: jest.fn(),
    refreshing: false,
    data: [
      {
        id: 123,
        title: 'enen',
        status: 'available',
        price: 3,
        currency: 'TWD'
      }
    ],
    onUpdateProductStatus: jest.fn(),
    onReachedEnd: jest.fn(),
    extraState: {},
    isChangingStatus: false,
    changingProductId: null
  };
  test('renders correctly', () => {
    const snapshot = renderer.create(<InventoryList {...props} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });
  const propsIsChanging = {
    onRefresh: jest.fn(),
    refreshing: false,
    data: [
      {
        id: 123,
        title: 'enen',
        status: 'available',
        price: 3,
        currency: 'TWD'
      }
    ],
    onUpdateProductStatus: jest.fn(),
    onReachedEnd: jest.fn(),
    extraState: {},
    isChangingStatus: true,
    changingProductId: 123
  };
  test('renders correctly', () => {
    const snapshot = renderer.create(<InventoryList {...propsIsChanging} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
