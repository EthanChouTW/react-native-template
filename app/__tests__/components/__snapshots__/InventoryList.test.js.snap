// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileMenuComponent renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "backgroundColor": "#E5E5E5",
      "paddingHorizontal": 14,
      "paddingVertical": 8,
    }
  }
  data={
    Array [
      Object {
        "currency": "TWD",
        "id": 123,
        "price": 3,
        "status": "available",
        "title": "enen",
      },
    ]
  }
  disableVirtualization={false}
  extraData={undefined}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={undefined}
  onEndReachedThreshold={5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <RCTRefreshControl />
  <>
    <
      onLayout={[Function]}
    >
      <
        style={
          Object {
            "backgroundColor": "white",
            "elevation": 2,
            "justifyContent": "center",
            "margin": 5,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.6,
            "shadowRadius": 1,
          }
        }
      >
        <
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "paddingHorizontal": 10,
              "paddingTop": 10,
            }
          }
        >
          enen
        </>
        <
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#E5E5E5",
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            }
          }
        >
          NT$3
        </>
        <
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#E5E5E5",
              "borderTopWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
              "paddingVertical": 20,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#DA3944",
                "textAlign": "center",
              }
            }
          >
            LBL_OUT_OF_STOCK
          </>
        </>
      </>
    </>
  </>
</RCTScrollView>
`;

exports[`UserProfileMenuComponent renders correctly 2`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "backgroundColor": "#E5E5E5",
      "paddingHorizontal": 14,
      "paddingVertical": 8,
    }
  }
  data={
    Array [
      Object {
        "currency": "TWD",
        "id": 123,
        "price": 3,
        "status": "available",
        "title": "enen",
      },
    ]
  }
  disableVirtualization={false}
  extraData={undefined}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={undefined}
  onEndReachedThreshold={5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  refreshing={false}
  removeClippedSubviews={false}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <RCTRefreshControl />
  <>
    <
      onLayout={[Function]}
    >
      <
        style={
          Object {
            "backgroundColor": "white",
            "elevation": 2,
            "justifyContent": "center",
            "margin": 5,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.6,
            "shadowRadius": 1,
          }
        }
      >
        <
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "paddingHorizontal": 10,
              "paddingTop": 10,
            }
          }
        >
          enen
        </>
        <
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#E5E5E5",
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            }
          }
        >
          NT$3
        </>
        <
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#E5E5E5",
              "borderTopWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
              "paddingVertical": 20,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <
            animating={true}
            color="rgba(43, 214, 101, 1)"
            hidesWhenStopped={true}
            size="small"
          />
        </>
      </>
    </>
  </>
</RCTScrollView>
`;
